Use constructor with parameter during serialization if possible

Jannify opened the pull request.

Jannify committed:
Implementation of using parameter constructors

Jannify committed:
Little performance increase (probably)

Jannify committed:
Fixed that properties without a setter wouldn't bet picked up

Jannify committed:
Performance boost: 2 ILs per serialized member for parameter constructor

Jannify committed:
Reworked ordering process of parameters

Jannify committed:
Rework Member sorting to take the possible constructor into account

SpaceMonkeyy86 commented on src/BinaryPack/Extensions/System/TypeExtensions.cs:
```diff
@@ -35,9 +35,8 @@ public static IReadOnlyCollection<MemberInfo> GetSerializableMembers(this Type t
-                       member is PropertyInfo propertyInfo && propertyInfo.CanRead && propertyInfo.CanWrite
+                       member is PropertyInfo propertyInfo && propertyInfo.CanRead
                        && propertyInfo.GetIndexParameters().Length == 0)
-                orderby member.Name
```
Why does this line need to be removed?

Jannify replied:
Moved to L47

SpaceMonkeyy86 commented on src/BinaryPack/Extensions/System/TypeExtensions.cs:
```diff
@@ -56,17 +55,65 @@ orderby member.Name
+            // Checking for constructor with no parameter and even any constructors first
+            if (type.GetConstructor(Type.EmptyTypes) != null || type.GetConstructors().Length == 0)
+            {
+                return members.OrderBy(member => member.Name);
```
If a parameterless constructor will be used for deserialization, the read-only properties cannot be serialized and should be filtered out here.

SpaceMonkeyy86 commented on src/BinaryPack/Extensions/System/TypeExtensions.cs:
```diff
@@ -56,17 +55,65 @@ orderby member.Name
+                }
+            }
+            
+            throw new NullReferenceException($"The type {type} has no constructor with parameters named like all Members.");
```
NullReferenceException is misleading here. Something like InvalidOperationException would be more appropriate for the issue.

SpaceMonkeyy86 commented on src/BinaryPack/Serialization/Processors/ObjectProcessor{T}.cs:
```diff
@@ -256,9 +257,26 @@ protected override void EmitDeserializer(ILGenerator il)
+
+                    if (parameterConstructor == null)
+                    {
+                        throw new NullReferenceException("The given object has neither an parameterless constructor nor a constructor with parameters named like all serialized fields/properties.");
```
A better exception type should be used here (see the comment in TypeExtensions.cs)

SpaceMonkeyy86 commented on src/BinaryPack/Extensions/System.Reflection.Emit/ILGeneratorExtensions.cs:
```diff
@@ -199,6 +207,16 @@ public static void EmitLoadLocalAddress(this ILGenerator il, int index)
+        
+        /// <inheritdoc cref="EmitLoadLocalForMemberWrite{T}(System.Reflection.Emit.ILGenerator,T,System.Reflection.MemberInfo)"/>
+        /// /// <param name="shouldSkip">Should this method be completely skipped</param>
+        public static void EmitLoadLocalForMemberWrite<T>(this ILGenerator il, T index, MemberInfo member, bool shouldSkip) where T : Enum
```
These method names should be changed to reflect their conditionality.

Jannify committed:
Filter out setterless properties for parameterless constructors

Jannify committed:
Changing some exceptions to a more fitting one

Jannify committed:
Moved custom methods in ILGeneratorExtensions to only class using it

Jannify committed:
Update gitignore for Rider files

Jannify committed:
Update gitignore for Rider files

SpaceMonkeyy86 commented on src/BinaryPack/Extensions/System/TypeExtensions.cs:
```diff
@@ -56,17 +55,66 @@ orderby member.Name
+            
+            foreach (ConstructorInfo constructor in type.GetConstructors())
+            {
+                if (constructor.GetParameters().Length != memberInfos.Count())
```
This rejects constructors that should (in theory) be valid because `memberInfos` also contains inherited members. The base `Packet` class has two members that never show up in constructors for derived types, so the program will treat those as invalid. Ideally, it should be able to recognize partial constructors and fill in the remaining properties using setters.

Alternatively, these properties are normally set to default values in the constructors, so another approach would be to create an attribute that causes the code finding parameter constructors to ignore certain properties.

Jannify replied:
Shouldn't the members you mentioned in `Packet.cs` (which shouldn't be serialized in the first place) marked with `IgnoredMemberAttribute`?

The other problem is with types that set inherited members in the constructor without a corresponding parameter. The only option I see to solve this is to have an attribute which forces to use a given constructor. Are you fine with this approach?

SpaceMonkeyy86 commented on src/BinaryPack/Extensions/System/TypeExtensions.cs:
```diff
@@ -56,17 +55,66 @@ orderby member.Name
+                }
+            }
+            
+            throw new InvalidOperationException($"The type {type} has no constructor with parameters named like all Members.");
```
```suggestion
            throw new InvalidOperationException($"The type {type} has no constructor with parameters corresponding to all members.");
```

Jannify committed:
Change exception text in TypeExtensions.cs

Jannify committed:
Added ForceUseConstructor attribute

This gives the option to use a constructor for serialization while parameter and member count are not equal.

SpaceMonkeyy86 committed:
Merge stuff

SpaceMonkeyy86 committed:
Fix a couple filters for serializable member search

SpaceMonkeyy86 commented:
It would also be good to have an `IgnoreConstructor` attribute for when an empty constructor needs to be present for another serializer.

Jannify committed:
Added IgnoreConstructorAttribute

SpaceMonkeyy86 committed:
Fix error messages

SpaceMonkeyy86 reviewed (approved):
Looks good, in the future we may consider revising to explicitly throw an error on certain cases, but I think this is fine for now.
